src/state.c

```diff
@@ -0,0 +1,60 @@
+#include <cstddef>
+#include <cstdint>
+#include <stdint.h>
+#include <stdlib.h>
+
+#include "./hash_table.h"
+#include "./state.h"
+#include "/usr/include/openssl/sha.h"
+
+client_state new_state(void) {
+  client_state new_state;
+  new_state.file_descriptors = make_table();
+  new_state.ports = make_table();
+  new_state.pieces_have = make_table();
+  new_state.pieces_want = make_table();
+  return new_state;
+}
+
+void dealloc_state(client_state *state) {
+  hash_dealloc(&state->file_descriptors);
+  hash_dealloc(&state->ports);
+  hash_dealloc(&state->pieces_have);
+  hash_dealloc(&state->pieces_want);
+}
+
+void add_piece_have(client_state *state, void *piece, size_t piece_size) {
```

Comment:
In `add_piece_have`, `remove_piece_have`, `add_piece_want`, `remove_piece_want`, `add_file_descriptor`, `remove_file_descriptor`, `add_port`, and `remove_port`, the variable `state` is a pointer to a `client_state`, not a client state itself.

E.G. `&state->pieces_have` instead of `state.have`